README.md
[![Build Status](https://travis-ci.org/Valloric/hamcrest2-rust.svg?branch=master)](https://travis-ci.org/Valloric/hamcrest2-rust)

# Hamcrest2

A port of [Hamcrest](http://hamcrest.org/) to [Rust](http://rust-lang.org).
Fork of original hamcrest-rust (which is unmaintained) with extra matchers,
better docs etc.

## Installing

To use Hamcrest, add this to your `Cargo.toml`:

```
[dev-dependencies]
hamcrest2 = "*"
```

And this to your crate root:

```rust
#[cfg(test)] #[macro_use] extern crate hamcrest2;
```

After a quick `cargo build`, you should be good to go!

## Usage

Hamcrest2 supports a number of matchers. The easiest way is to just `use` them all like this:

```rust
use hamcrest2::prelude::*;
```

If you want to be more selective make sure that you also import the `HamcrestMatcher` trait.

## General Matchers

### eq, not

```rust
assert_that!(1, eq(1));  // also equal_to()
assert_that!(1, not(eq(2)));
```

### compared_to

```rust
assert_that!(1, lt(2));   // also less_than()
assert_that!(1, leq(1));  // also less_than_or_equal_to()
assert_that!(2, gt(1));   // also greater_than()
assert_that!(2, geq(2));  // also greater_than_or_equal_to()
```

### type_of

```rust
assert_that!(123usize, type_of::<usize>());
assert_that!("test", type_of::<&str>());
```

### matches_regex

```rust
assert_that!("1234", matches_regex(r"\d"));
assert_that!("abc", does_not(match_regex(r"\d")));
```

## Numerical Matchers

### close_to

```rust
assert_that!(1e-40f32, close_to(0.0, 0.01));
assert_that!(1e-40f32, not(close_to(0.0, 0.000001)));
```

## Filesystem Matchers

### path_exists, file_exists, dir_exists

```rust
let path = Path::new("./README.md");
assert_that!(path, path_exists());
assert_that!(path, file_exists());
assert_that!(path, not(dir_exists()));
```

## Option and Result

### has

```rust
let var: Option<i8> = Some(5);
assert_that!(var, has(5));

let var: Result<i8, String> = Ok(5);
assert_that!(var, has(5));
```

### ok
```rust
let var: Result<i8, String> = Ok(5);
assert_that!(var, ok());

assert_that!(Ok(5), ok::<i8, String>());

let var: Result<i8, String> = Err("bad!".to_string());
assert_that!(var, not(ok()));
```

### err

```rust
let var: Result<i8, String> = Err("bad!".to_string());
assert_that!(var, err());

assert_that!(Err("bad!".to_string()), err::<i8, String>());

let var: Result<i8, String> = Ok(5);
assert_that!(var, not(err()));
```

### some

```rust
let var: Option<i8> = Some(5);
assert_that!(var, some());

assert_that!(Some(1), some::<u8>());

let var: Option<i8> = None;
assert_that!(var, not(some()));
```

### none

```rust
let var: Option<i8> = None;
assert_that!(var, none());

assert_that!(None, none::<u8>());
assert_that!(Some(1), not(none::<u8>()));
```

## Collection Matchers

### contains, contains exactly, contains in order

```rust
assert_that!(&vec!(1, 2, 3), contains(vec!(1, 2)));
assert_that!(&vec!(1, 2, 3), not(contains(vec!(4i))));

assert_that!(&vec!(1, 2, 3), contains(vec!(1, 2, 3)).exactly());
assert_that!(&vec!(1, 2, 3), not(contains(vec!(1, 2)).exactly()));

assert_that!(&vec!(1, 2, 3), contains(vec!(1, 2)).in_order());
assert_that!(&vec!(1, 2, 3), not(contains(vec!(1, 3)).in_order()));
```

## len
```rust
assert_that!(&vec!(1, 2, 3), len(3));
assert_that!(&vec!(1, 2, 3), not(len(4)));
```

## empty
```rust
assert_that!(&Vec::<i32>::new(), empty());
assert_that!(&vec![1, 2, 3], not(empty()));
```

## Compound Matchers

### all

```rust
assert_that!(4, all!(lt(5), gt(3)));  // also and!()
assert_that!(
    &vec![1, 2, 3],
    all!(contains(vec![1, 2]), not(contains(vec![4])))
);
```

### any

```rust
assert_that!(4, any!(less_than(2), greater_than(3)));  // also or!()
assert_that!(
    &vec![1, 2, 3],
    any!(contains(vec![1, 2, 5]), not(contains(vec![4])))
);
```

## Misc Matchers

### is(bool)

```rust
assert_that!(true, is(true));
assert_that!(false, is(false));
```

### anything

```rust
assert_that!(42, anything());
assert_that!("test", is(anything()));
```

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the
work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
## 0.2.6 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.2.4...0.2.6)

* The previous version introduced support for reference arguments and this 
broke `equal_to` for slices. Thus, automatic ref argument support for 
`equal_to` has been reverted.

## 0.2.4 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.2.3...0.2.4)

* `contains` matcher is now generic and supports both collection and single item
  arguments; thus, both `contains(vec![5])` and `contains(5)` work
* Almost all matchers now support reference arguments as well!

## 0.2.3 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.2.2...0.2.3)

* `contains`, `empty` and `len` matchers now work for slices, not just vectors

## 0.2.2 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.2.1...0.2.2)

* Added the `empty` matcher

## 0.2.1 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.2.0...0.2.1)

* Better message for expected output for `ok`, `err`, `some` and `none` matchers

## 0.2.0 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.1.6...0.2.0)

* Created real crate docs with doctests
* added `ok` and `err` matchers
* `of_len!` is now `len!`
* `any_of!` is now `any!`
* `all_of!` is now `all!`
* `existing_path` is now `path_exists`
* `existing_file` is now `file_exists`
* `existing_dir` is now `dir_exists`

## 0.1.6 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.1.5...0.1.6)

* Shorter names for common matchers:
    * `eq` for `equal_to`
    * `lt` for `less_than`
    * `gt` for `greater_than`
    * similarly, `geq`, `leq` etc
* Restructured examples in README to reduce verbosity
* Added `some` matcher
* Added `has` matcher (like `contains` but for `Option` and `Result`!)

## 0.1.5 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.1.4...0.1.5)

* Implemented matcher trait for boolean values, #48

## 0.1.4 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.1.3...0.1.4)

* Logical matchers `all_of`, `any_of`, comparison matchers `type_of`, `anything`, #47

## 0.1.3 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.1.2...0.1.3)

* Comparison matchers `less_than`, `less_than_or_equal_to`, `greater_than`, `greater_than_or_equal_to`. #43
* `in_order` option for `contains`. #44

## 0.1.2 [☰](https://github.com/Valloric/hamcrest2-rust/compare/0.1.1...0.1.2)

* Added the `assert_that!` macro. It produces better error messages (with correct file and line
  number).
* Deprecated the `assert_that` function.
* Improvements to `Cargo.toml` (by @killercup)

## 0.1.1 [☰](https://github.com/Valloric/hamcrest2-rust/compare/a9f18681c64e3126ef6ccbd68ec2a5b39fe5b58b...0.1.1)

* Licensing change. The crate is now dual licensed under the MIT and Apache 2 licenses.
* Adds the `prelude` submodule to simplify inclusion of all matchers.
* `matches_regex` matcher


src/lib.rs
// Copyright 2014 Carl Lerche, Oliver Mader, Alex Crichton, Thiago Pontes,
//                Yehuda Katz
// Copyright 2015 Carl Lerche, Oliver Mader
// Copyright 2016 Urban Hafner
// Copyright 2018 Val Markovic
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! ## Usage
//!
//! Hamcrest2 supports a number of matchers. The easiest way is to just `use`
//! them all like this:
//!
//! ```
//! use hamcrest2::prelude::*;
//! ```
//!
//! If you want to be more selective make sure that you also import the
//! `HamcrestMatcher` trait.
//!
//! ## General Matchers
//!
//! ### eq, not
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(1, eq(1));  // also equal_to()
//! assert_that!(1, not(eq(2)));
//! ```
//!
//! ### compared\_to
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(1, lt(2));   // also less_than()
//! assert_that!(1, leq(1));  // also less_than_or_equal_to()
//! assert_that!(2, gt(1));   // also greater_than()
//! assert_that!(2, geq(2));  // also greater_than_or_equal_to()
//! ```
//!
//! ### type_of
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(123usize, type_of::<usize>());
//! assert_that!("test", type_of::<&str>());
//! ```
//!
//! ### matches_regex
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!("1234", matches_regex(r"\d"));
//! assert_that!("abc", does_not(match_regex(r"\d")));
//! ```
//!
//! ## Numerical Matchers
//!
//! ### close_to
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(1e-40f32, close_to(0.0, 0.01));
//! assert_that!(1e-40f32, not(close_to(0.0, 0.000001)));
//! ```
//!
//! ## Filesystem Matchers
//!
//! ### path_exists, file_exists, dir_exists
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! # pub use std::path::Path;
//! let path = Path::new("./README.md");
//! assert_that!(path, path_exists());
//! assert_that!(path, file_exists());
//! assert_that!(path, not(dir_exists()));
//! ```
//!
//! ## Option and Result
//!
//! ### has
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! let var: Option<i8> = Some(5);
//! assert_that!(var, has(5));
//!
//! let var: Result<i8, String> = Ok(5);
//! assert_that!(var, has(5));
//! ```
//! ### ok
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! let var: Result<i8, String> = Ok(5);
//! assert_that!(var, ok());
//!
//! assert_that!(Ok(5), ok::<i8, String>());
//!
//! let var: Result<i8, String> = Err("bad!".to_string());
//! assert_that!(var, not(ok()));
//! ```
//!
//! ### err
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! let var: Result<i8, String> = Err("bad!".to_string());
//! assert_that!(var, err());
//!
//! assert_that!(Err("bad!".to_string()), err::<i8, String>());
//!
//! let var: Result<i8, String> = Ok(5);
//! assert_that!(var, not(err()));
//! ```
//!
//! ### some
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! let var: Option<i8> = Some(5);
//! assert_that!(var, some());
//!
//! assert_that!(Some(1), some::<u8>());
//!
//! let var: Option<i8> = None;
//! assert_that!(var, not(some()));
//! ```
//!
//! ### none
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! let var: Option<i8> = None;
//! assert_that!(var, none());
//!
//! assert_that!(None, none::<u8>());
//! assert_that!(Some(1), not(none::<u8>()));
//! ```
//!
//! ## Collection Matchers
//!
//! ### contains, contains exactly, contains in order
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(&vec!(1, 2, 3), contains(vec!(1, 2)));
//! assert_that!(&vec!(1, 2, 3), not(contains(vec!(4))));
//!
//! assert_that!(&vec!(1, 2, 3), contains(vec!(1, 2, 3)).exactly());
//! assert_that!(&vec!(1, 2, 3), not(contains(vec!(1, 2)).exactly()));
//!
//! assert_that!(&vec!(1, 2, 3), contains(vec!(1, 2)).in_order());
//! assert_that!(&vec!(1, 2, 3), not(contains(vec!(1, 3)).in_order()));
//! ```
//!
//! ## len
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(&vec!(1, 2, 3), len(3));
//! assert_that!(&vec!(1, 2, 3), not(len(4)));
//! ```
//!
//! ## empty
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(&Vec::<i32>::new(), empty());
//! assert_that!(&vec![1, 2, 3], not(empty()));
//! ```
//!
//! ## Compound Matchers
//!
//! ### all
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(4, all!(lt(5), gt(3)));  // also and!()
//! assert_that!(
//!     &vec![1, 2, 3],
//!     all!(contains(vec![1, 2]), not(contains(vec![4])))
//! );
//! ```
//!
//! ### any
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(4, any!(less_than(2), greater_than(3)));  // also or!()
//! assert_that!(
//!     &vec![1, 2, 3],
//!     any!(contains(vec![1, 2, 5]), not(contains(vec![4])))
//! );
//! ```
//!
//! ## Misc Matchers
//!
//! ### is(bool)
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(true, is(true));
//! assert_that!(false, is(false));
//! ```
//!
//! ### anything
//!
//! ```
//! # #[macro_use] extern crate hamcrest2;
//! # use hamcrest2::prelude::*;
//! assert_that!(42, anything());
//! assert_that!("test", is(anything()));
//! ```

extern crate num;
extern crate regex;

pub use prelude::*;

#[macro_export]
macro_rules! assert_that {
  ($actual:expr, $matcher:expr) => {{
    // The separate statement is necessary to keep the compiler happy.
    let m = $matcher;
    match m.matches($actual) {
      Ok(_) => {}
      Err(mismatch) => {
        // The panic macro produces the correct file and line number
        // when used in a macro like this, i.e. it's the line where
        // the macro was originally written.
        panic!("\nExpected: {}\n    but: {}", m, mismatch);
      }
    }
  }};
}

pub mod core;
pub mod matchers;
mod utils;
pub mod prelude {
  #[allow(deprecated)]
  pub use core::assert_that;
  pub use core::Matcher as HamcrestMatcher;
  pub use matchers::all::all;
  #[deprecated(since = "0.2.0", note = "Use all() instead")]
  pub use matchers::all::all as all_of;
  #[deprecated(since = "0.2.0", note = "Use all() instead")]
  pub use matchers::all::all as and;
  pub use matchers::any::any;
  #[deprecated(since = "0.2.0", note = "Use any() instead")]
  pub use matchers::any::any as any_of;
  #[deprecated(since = "0.2.0", note = "Use any() instead")]
  pub use matchers::any::any as or;
  pub use matchers::anything::anything;
  pub use matchers::close_to::close_to;
  pub use matchers::compared_to::greater_than;
  pub use matchers::compared_to::greater_than as gt;
  pub use matchers::compared_to::greater_than_or_equal_to;
  pub use matchers::compared_to::greater_than_or_equal_to as geq;
  pub use matchers::compared_to::less_than;
  pub use matchers::compared_to::less_than as lt;
  pub use matchers::compared_to::less_than_or_equal_to;
  pub use matchers::compared_to::less_than_or_equal_to as leq;
  pub use matchers::contains::contains;
  pub use matchers::empty::empty;
  pub use matchers::equal_to::equal_to;
  pub use matchers::equal_to::equal_to as eq;
  pub use matchers::err::err;
  pub use matchers::has::has;
  pub use matchers::is::is;
  pub use matchers::is::is_not as does_not;
  pub use matchers::is::is_not as not;
  pub use matchers::is::is_not;
  pub use matchers::len::len;
  #[deprecated(since = "0.2.0", note = "Use len() instead")]
  pub use matchers::len::len as of_len;
  pub use matchers::none::none;
  pub use matchers::ok::ok;
  pub use matchers::path_exists::dir_exists;
  #[deprecated(since = "0.2.0", note = "Use dir_exists() instead")]
  pub use matchers::path_exists::dir_exists as existing_dir;
  pub use matchers::path_exists::file_exists;
  #[deprecated(since = "0.2.0", note = "Use file_exists() instead")]
  pub use matchers::path_exists::file_exists as existing_file;
  pub use matchers::path_exists::path_exists;
  #[deprecated(since = "0.2.0", note = "Use path_exists() instead")]
  pub use matchers::path_exists::path_exists as existing_path;
  pub use matchers::regex::matches_regex as match_regex;
  pub use matchers::regex::matches_regex;
  pub use matchers::some::some;
  pub use matchers::type_of::type_of;
}
